Re-Enable PhaseTracker Tracking and Block Tracking

gabizou opened the pull request:
Work in progress

# What's enabled

- Wrapping PhaseState switches for Entity Ticks, BlockEntity ticks, Block ticks
- A majority of Timings for the same types
- PhaseTracker specific logic code that requires mixins are all in their own mixin module

# What needs to be done
- [x] Block Tracking
- [x] TileEntity tracking
- [ ] Block Physics rewiring
- [ ] Block Event throws

# Current state of things:

## A new way to record Transactions

While evaluating the mod compatibility and complexity introduced by the previous iteration of MultiBlockCaptureSupplier, @dualspiral and I elected to merely attach a _recorder_ to log what changes a `World` was going through. The end result is a **new** system that recycles the core concept of the `PhaseTracker` keeping track of contexts, into what we're calling `TransactionalCaptureSupplier`. 

![image](https://user-images.githubusercontent.com/1203877/89111976-92c2e480-d411-11ea-9440-45d0076d08ab.png)
<subscript>meme credit by @Faithcaio</subscript>


## Bugs
- Client Notifications stopped working for all pipeline based interactions

gabizou committed:
Start turning on some tracking

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Enable PhaseTracker wrapping around some updates

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou commented on src/mixins/java/org/spongepowered/common/mixin/tracker/world/chunk/ChunkMixin_Tracker.java:
```diff
@@ -421,24 +439,25 @@ public BlockState setBlockState(final BlockPos pos, final BlockState state, fina
             this.tileEntities.get(pos).remove();
         }
         added.validate();
-        ((ActiveChunkReferantBridge) added).bridge$setActiveChunk((ChunkBridge) this);
+        ((ActiveChunkReferantBridge) added).bridge$setActiveChunk(this);
         this.tileEntities.put(pos, added);
     }
 
     @Inject(
             method = "addTileEntity(Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/tileentity/TileEntity;)V",
             at = @At(value = "INVOKE", target = "Lnet/minecraft/tileentity/TileEntity;validate()V"))
     private void tracker$SetActiveChunkOnTileEntityAdd(final BlockPos pos, final TileEntity tileEntityIn, final CallbackInfo ci) {
-        ((ActiveChunkReferantBridge) tileEntityIn).bridge$setActiveChunk((ChunkBridge) this);
+        ((ActiveChunkReferantBridge) tileEntityIn).bridge$setActiveChunk(this);
         // Make sure to set creator/notifier for TE if any chunk data exists
         // Failure to do this during chunk load will cause TE's to not have proper user tracking
-        ((CreatorTrackedBridge) tileEntityIn).tracked$setTrackedUUID(PlayerTracker.Type.CREATOR, ((ChunkBridge) this).bridge$getBlockCreatorUUID(pos).orElse(null));
-        ((CreatorTrackedBridge) tileEntityIn).tracked$setTrackedUUID(PlayerTracker.Type.NOTIFIER, ((ChunkBridge) this).bridge$getBlockNotifierUUID(pos).orElse(null));
+        // TODO - Reimplement player uuid tracking.
+        ((CreatorTrackedBridge) tileEntityIn).tracked$setTrackedUUID(PlayerTracker.Type.CREATOR, null);
+        ((CreatorTrackedBridge) tileEntityIn).tracked$setTrackedUUID(PlayerTracker.Type.NOTIFIER, null);
```
@Zidane this is the bit that if you revert the changes I made here, this will cause the client to just hang itself on loading a world in single player.

gabizou committed:
Try enabling World player tracking for positions

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Zidane commented on src/mixins/java/org/spongepowered/common/mixin/core/world/WorldSettingsMixin.java:
```diff
@@ -71,9 +71,10 @@
+    private void impl$reAssignValuesFromIncomingInfo(final WorldInfo info, final CallbackInfo ci) {
         final WorldProperties properties = (WorldProperties) info;
         if (((WorldInfoBridge) info).bridge$isValid()) {
+            this.dimensionType = (SpongeDimensionType) properties.getDimensionType();
```
Try it twice, it'll work the second time?

gabizou committed:
Implmement BlockChangeFlag.Factory

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Try enabling only single block tracking

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Add BlockStateBridge in lieu of SpongeImplHooks

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Update setBlockState logic sans tracking in effect

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Start skeleton implementation of a new Transactor recorder pipeline

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Pipelines are starting to get built

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Complete WorldPipeline for block transacting

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Remove unused sections of BlockTransaction. Enable warning on Chunk direct block access

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Restrict who and when can call the tracked block change pipeline

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Small fixes to get the game launching again and the pipeline actually used

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Add TileEntity related Pipelines

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Remove legacy BlockTransaction and MultiBlockCaptureSupplier

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Inline PhaseTracker.setBlockState to ServerWorldMixin_Tracker

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Inline neighbor notification execution back into ServerWorldMixin_Tracker

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Default implement TrackedWorldBridge.bridge$createSnapshot

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Migrate neighbor notification override logic to Tracker module

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Eliminate redundant methods from IPhaseState related to legacy block capturing

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Remove singular legacy transaction capturing from PhaseContext

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Make the Sponge physics mask constant legible/properly masked

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Rewrite SpongeLocatableBlock to use a Supplier for the world

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Append effects to parents when logging children transactions

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Update PacketThreadUtilMixin to tracker module and enable packet states

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Implement new neighbor notification effect transaction recording

This has the side effect of eliminating the need for a NeighborNotificationState since we no longer
switch in to new PhaseContexts for neighbor events. NeighborNotification as a transaction will suit
enough of the needs to be able to explicitly record related transactions while reducing the overall
need to enter a new stack element to the JVM stack for processing.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Move EffectTransactor out to its own class

This allows for effect transactors to eliminate themselves from
the pipe if there are no transactions being recorded during their
effects process.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Implement Event generation from BlockTransactions
